backtrace-library/src/androidTest/java/backtraceio/library/BacktraceClientMetricsTest.java

```diff
@@ -0,0 +1,956 @@
+        // We will always have startup unique event GUID
+        assertEquals(1, backtraceClient.getUniqueEvents().size());
+
+        backtraceClient.sendMetrics();
+
+        try {
+            waiter.await(1000);
+        } catch (Exception e) {
+            fail(e.toString());
+        }
+
+        assertNotEquals(0, mockRequestHandler.lastEventPayloadJson.length());
+        assertEquals(1, mockRequestHandler.numAttempts);
+        assertEquals(0, backtraceClient.getSummedEvents().size());
+        // We will always have startup unique event GUID
+        assertEquals(1, backtraceClient.getUniqueEvents().size());
+    }
+
+    @Test
+    public void uploadUniqueEventsManual() {
+        final Waiter waiter = new Waiter();
+
+        backtraceClient.enableMetrics(new BacktraceMetricsSettings(universeName, token, defaultBaseUrl, 0));
+        MockRequestHandler mockRequestHandler = new MockRequestHandler();
+        backtraceClient.setUniqueEventsRequestHandler(mockRequestHandler);
+
+        backtraceClient.setUniqueEventsOnServerResponse(new EventsOnServerResponseEventListener() {
+            @Override
+            public void onEvent(EventsResult result) {
+                //Assert.assertEquals(resultMessage, result.message);
+                assertEquals(0, result.getEventsPayload().getDroppedEvents());
+                String eventsPayloadJson = BacktraceSerializeHelper.toJson(result.getEventsPayload());
+                assertNotEquals(0, eventsPayloadJson.length());
+                assertEquals(BacktraceResultStatus.Ok, result.status);
+                waiter.resume();
+            }
+        });
+
+        backtraceClient.addUniqueEvent(uniqueAttributeName[0]);
+        backtraceClient.sendMetrics();
+
+        // We will always have startup unique event GUID
+        assertEquals(2, backtraceClient.getUniqueEvents().size());
+
+        try {
+            waiter.await(1000);
+        } catch (Exception e) {
+            fail(e.toString());
+        }
+
+        assertNotEquals(0, mockRequestHandler.lastEventPayloadJson.length());
+        assertEquals(1, mockRequestHandler.numAttempts);
+        // We will always have startup unique event GUID
+        assertEquals(2, backtraceClient.getUniqueEvents().size());
+        assertEquals(0, backtraceClient.getSummedEvents().size());
+    }
+
+    @Test
+    public void doNotUploadWhenNoEventsAvailable() {
+        backtraceClient.enableMetrics(new BacktraceMetricsSettings(universeName, token, defaultBaseUrl, 0));
+
+        // Clear startup event
+        ConcurrentLinkedDeque<UniqueEvent> uniqueEvents = backtraceClient.getUniqueEvents();
+        uniqueEvents.clear();
+
+        MockRequestHandler mockRequestHandler = new MockRequestHandler();
+        backtraceClient.setUniqueEventsRequestHandler(mockRequestHandler);
+
+        // When no events in queue request handler should not be called
+        backtraceClient.sendMetrics();
+        try {
+            sleep(1000);
```

Comment:
Isn't there another way without explicit numbered ms waits?

Reply:
The waiter seems cool - I guess you can't wait for no interactions?

Reply:
I will mark this as resolved @vlussenburg